### ISSUE TITLE
Support for Cloudflare R2 Object Storage

### ISSUE THREAD

**Issue by schopenhauer** [rclone/rclone#5642]
As announced today by Cloudflare.

https://blog.cloudflare.com/introducing-r2-object-storage/

----

### Update from @ncw 

The aim for this ticket is to make `R2` a fully supported S3 provider.

Some initial progress on this has been made now that the beta is open. In particular #5422 will be required since R2 doesn't support presigned URLs.

With #5422 it works reasonably well with a config like this.

```
[r2]
type = s3
provider = Other
access_key_id = ACCESSKEY
secret_access_key = SECRETKEY
endpoint = https://YOURCLOUDFLAREID.r2.cloudflarestorage.com
region = auto
list_url_encode = true
```

**Comment by ncw** [comment]
I've requested an account. If anyone at Cloudflare is reading this and would like to send me a beta test account please contact me at nick@craig-wood.com !

From my understanding of the docs this should be an S3 compatible remote, but we'll see...

**Comment by greg-mckeon** [comment]
someone at Cloudflare is reading this :).  We'll get you an account when we're a bit more opened up - please ping me here if that doesn't happen.

**Comment by rsignell-usgs** [comment]
Did this happen? Would love to see R2 support!

**Comment by ncw** [comment]
Here is the latest test for Cloudflare. This has the Cloudflare provider integrated now.

[v1.59.0-beta.6120.fb1caaaef.fix-5422-s3-putobject](https://beta.rclone.org/branch/fix-5422-s3-putobject/v1.59.0-beta.6120.fb1caaaef.fix-5422-s3-putobject/) on branch [fix-5422-s3-putobject](https://github.com/rclone/rclone/tree/fix-5422-s3-putobject) (uploaded in 15-30 mins)

Your config file should look like this

```
[r2]
type = s3
provider = Cloudflare
access_key_id = ACCESSKEY
secret_access_key = SECRETKEY
endpoint = https://YOURCLOUDFLAREID.r2.cloudflarestorage.com
region = auto
```

This passes very nearly all the integration tests so should be good for intensive testing.

**Comment by ncw** [comment]
Here is the latest update - changes from the last version are to do with error messages on upload failures. This now passes all the integration tests.

[v1.59.0-beta.6122.7a0cdbc45.fix-5422-s3-putobject](https://beta.rclone.org/branch/fix-5422-s3-putobject/v1.59.0-beta.6122.7a0cdbc45.fix-5422-s3-putobject/) on branch [fix-5422-s3-putobject](https://github.com/rclone/rclone/tree/fix-5422-s3-putobject) (uploaded in 15-30 mins)

**Comment by johnspurlock** [comment]
That beta build is working great for me in simple local -> r2, r2 -> local, and r2 -> r2 syncs.

Can't wait to see this in an official release build!

**Comment by Erisa** [comment]
Working great for me too, now that the issue have been cleared up on both rclone and R2's ends. Don't have any errors anymore in anything I try to do.

**Comment by lpellegr** [comment]
Works great too! Is there no chance to get a release for the beta announcement of R2 happening later Today?

**Comment by ncw** [comment]
This depends on #5422 which I've tested with lots of S3 backends now so should be good to merge.

I need to add docs, then I can merge this to master.  The changes in #5422 are too big for a point release so it will have to wait for v1.59 for a full release.

**Comment by lpellegr** [comment]
Thanks for your answer. Would be great if the latest beta include the changes because it is really easy to install:

> curl https://rclone.org/install.sh | sudo bash -s beta

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.59.

You can install rclone betas with `curl https://rclone.org/install.sh | sudo bash -s beta` also.

The help looks like this

### Cloudflare R2

[Cloudflare R2](https://blog.cloudflare.com/r2-open-beta/) Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.

Here is an example of making a Cloudflare R2 configuration. First run:

    rclone config

This will guide you through an interactive setup process.

```
No remotes found, make a new one?
n) New remote
s) Set configuration password
q) Quit config
n/s/q> n
name> r2
Option Storage.
Type of storage to configure.
Choose a number from below, or type in your own value.
...
XX / Amazon S3 Compliant Storage Providers including AWS, Alibaba, Ceph, China Mobile, Cloudflare, ArvanCloud, Digital Ocean, Dreamhost, IBM COS, Lyve Cloud, Minio, Netease, RackCorp, Scaleway, SeaweedFS, StackPath, Storj, Tencent COS and Wasabi
   \ (s3)
...
Storage> s3
Option provider.
Choose your S3 provider.
Choose a number from below, or type in your own value.
Press Enter to leave empty.
...
XX / Cloudflare R2 Storage
   \ (Cloudflare)
...
provider> Cloudflare
Option env_auth.
Get AWS credentials from runtime (environment variables or EC2/ECS meta data if no env vars).
Only applies if access_key_id and secret_access_key is blank.
Choose a number from below, or type in your own boolean value (true or false).
Press Enter for the default (false).
 1 / Enter AWS credentials in the next step.
   \ (false)
 2 / Get AWS credentials from the environment (env vars or IAM).
   \ (true)
env_auth> 1
Option access_key_id.
AWS Access Key ID.
Leave blank for anonymous access or runtime credentials.
Enter a value. Press Enter to leave empty.
access_key_id> ACCESS_KEY
Option secret_access_key.
AWS Secret Access Key (password).
Leave blank for anonymous access or runtime credentials.
Enter a value. Press Enter to leave empty.
secret_access_key> SECRET_ACCESS_KEY
Option region.
Region to connect to.
Choose a number from below, or type in your own value.
Press Enter to leave empty.
 1 / R2 buckets are automatically distributed across Cloudflare's data centers for low latency.
   \ (auto)
region> 1
Option endpoint.
Endpoint for S3 API.
Required when using an S3 clone.
Enter a value. Press Enter to leave empty.
endpoint> https://ACCOUNT_ID.r2.cloudflarestorage.com
Option acl.
Canned ACL used when creating buckets and storing or copying objects.
This ACL is used for creating objects and if bucket_acl isn't set, for creating buckets too.
For more info visit https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl
Note that this ACL is applied when server-side copying objects as S3
doesn't copy the ACL from the source but rather writes a fresh one.
Choose a number from below, or type in your own value.
Press Enter to leave empty.
   / Owner gets FULL_CONTROL.
 1 | No one else has access rights (default).
   \ (private)
...
acl> 1
Edit advanced config?
y) Yes
n) No (default)
y/n> n
--------------------
y) Yes this is OK (default)
e) Edit this remote
d) Delete this remote
y/e/d> y
```

This will leave your config looking something like:

```
[r2]
type = s3
provider = Cloudflare
access_key_id = ACCESS_KEY
secret_access_key = SECRET_ACCESS_KEY
region = auto
endpoint = https://ACCOUNT_ID.r2.cloudflarestorage.com
acl = private
```

Now run `rclone lsf r2:` to see your buckets and `rclone lsf
r2:bucket` to look within a bucket.

**Comment by bbhxwl** [comment]
![image](https://user-images.githubusercontent.com/26525599/168198486-889cbcd8-ce5e-4d0f-b25a-ddc42611bcb8.png)

**Comment by ncw** [comment]
@bbhxwl - try again now. This is an occasional glitch which I haven't got to the bottom of yet!

**Comment by stefanmaric** [comment]
I'm using the latest beta and the `rclone sync` command isn't behaving as I would expect: syncing a local folder to R2 takes the same as the initial upload.

```shell
$ rclone --version
rclone v1.59.0-beta.6163.6d342a3c5
- os/version: fedora 36 (64 bit)
- os/kernel: 5.17.9-300.fc36.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.18.2
- go/linking: static
- go/tags: none
```

Running the command with `-vv` I notice it is copying every file, even when running the same sync command on the same local folder (without any modifications) against the same bucket and prefix:

```shell
...
2022/05/25 13:19:53 DEBUG : assets/apple-touch-startup-image-750x1334.png: md5 = 11bd3dc093a9e1ec2222d503c56fbedf OK
2022/05/25 13:19:53 INFO  : assets/apple-touch-startup-image-750x1334.png: Copied (new)
2022/05/25 13:19:53 DEBUG : assets/apple-touch-startup-image-828x1792.png: md5 = c12f4079f5c33964d0b919be68eb8d18 OK
2022/05/25 13:19:53 INFO  : assets/apple-touch-startup-image-828x1792.png: Copied (new)
2022/05/25 13:19:53 DEBUG : assets/favicon-32x32.png: md5 = 8ece8020e7428b394833f1de2528d223 OK
2022/05/25 13:19:53 INFO  : assets/favicon-32x32.png: Copied (new)
2022/05/25 13:19:53 DEBUG : assets/favicon-16x16.png: md5 = 04ffe7abffecdc1a12c1f5e68d6ed7cb OK
2022/05/25 13:19:53 INFO  : assets/favicon-16x16.png: Copied (new)
...
```

I've tried different combinations of `--fast-list`, `--checksum` and `--s3-no-head` without success, it copies everything and always take roughly the same time (about 1min 20sec for ~30mb of web assets).

Based on R2 documentation it seems all write operations support the md5 metadata field so I would expect rclone to compare these: https://developers.cloudflare.com/r2/platform/s3-compatibility/api/

Am I missing something? :thinking:

**Comment by ncw** [comment]
@stefanmaric I don't see this in a quick test

```
$ rclone sync -P /tmp/filez/ r2:rclone-sync-test
Transferred:        4.736 KiB / 4.736 KiB, 100%, 188 B/s, ETA 0s
Transferred:          100 / 100, 100% <--- 100 files transferred
Elapsed time:        27.8s
```

```
$ rclone sync -P /tmp/filez/ r2:rclone-sync-test
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:               100 / 100, 100% <---- 100 files checked and not transferred
Elapsed time:         3.4s
```



> ```shell
> 2022/05/25 13:19:53 INFO  : assets/favicon-16x16.png: Copied (new)
> ```

The `Copied (new)` part here means that rclone didn't find the existing file at all so are the files going to the same place.

Please provide a way for me to reproduce - thank you :-)

**Comment by Lusitaniae** [comment]
It's not really clear to me where to find key and key secret

CF only generates one value (the token/secret)

**Comment by Erisa** [comment]
> It's not really clear to me where to find key id and key secret
> 
> CF only generates one value (the token/secret)

Rather than creating API tokens on your Cloudflare profile, you need to make them on the R2 section of the dasboard.

Either go R2 -> "Manage R2 API Tokens" on the dashboard
Or go direct to https://dash.cloudflare.com/?to=/:account/r2/api-tokens

**Comment by ribasushi** [comment]
@ncw this is awesome! Is there a reason that only `lsf` works and not `lsl`?

**Comment by ncw** [comment]
Recursive listings are currently broken for R2. I reported this yesterday and they told me a fix was imminent. In the meantime `--disable ListR` will work around.

**Comment by cvoadmin** [comment]
Can this run in a worker?

**Comment by Erisa** [comment]
> Can this run in a worker?

No, you can't run rclone in a Worker in any way.

When interacting with Cloudflare R2 from a Worker, you can use [the native R2 Worker bindings](https://developers.cloudflare.com/r2/runtime-apis/) or an S3 client like [aws4fetch](https://github.com/mhart/aws4fetch).

**Comment by ncw** [comment]
This is all merged into 1.59 now, so I'm going to close this. If anyone finds any problems, please open a new issue.
